### PR TITLE
Implement some reductions for string Series

### LINKED ISSUE #18588: Inconsistent behaviour with min method on object dtype columns
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({'col1': [0, 1, 2, 3], 'col2': ['a', 'b', None, 'd'], 'col3': ['e', 'f', None, 'h']})
df.min(axis=0, skipna=True)
```
#### Problem description

If I've read the documentation correctly I think min should return a series of the same length as the relevant axis of the dataframe, and especially with the skipna flag set as True (which is the default) NA values should be ignored in the calculation.

#### Output

col1    0
dtype: int64

#### Expected Output

col1    0
col2    a
col3    e
dtype: object

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.3
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #31746: New string data type aggregations (min, max, sum) work for DataFrames but not Series
```python
>>> s = pd.Series(['a', 'b'], dtype='string')
>>> s.max()
TypeError: Cannot perform reduction 'max' with string dtype
>>> df = s.to_frame()
>>> df.max()
0    b
dtype: object
```
#### Problem description

I assume this isn't supposed to work on DataFrames if it doesn't work on strings. min and sum do the same thing.

#### Expected Output

Consistent behavior for both dataframes and series

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.2.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0+untagged.1.gce8af21.dirty
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 45.1.0.post20200127
Cython           : 0.29.14
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.0
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.13
tables           : None
tabulate         : 0.8.3
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #31746, closes #18588
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/arrays/string_/test_string.py [hunk 1]
```diff
@@ -269,3 +267,31 @@ def test_value_counts_na():
...
+
+    result = getattr(arr, func)(skipna=skipna)
+    expected = getattr(ser, func)(skipna=skipna)
+
+    assert result == expected
+
+
+@pytest.mark.parametrize("func", ["min", "max"])
+@pytest.mark.parametrize("skipna", [True, False])
+def test_reduction_with_na(func, skipna):
```
Comment (jbrockmendel) [inline pandas/tests/arrays/string_/test_string.py]: any chance of this being parametrized over StringArray/Index/Series?  if so, these might belong in tests.reductions

Comment (dsaxton) [reply pandas/tests/arrays/string_/test_string.py]: As written right now I think this would only work with Series. Perhaps these methods should get implemented correctly for StringArray, and then delegate to those within _reduce instead of using the parent class methods (then test both StringArray and Series at once)?

**Comment by WillAyd** [comment]
Just as a counterpoint do we really want to support all of these? `sum` in particular is strange to me to support on a string dtype

**Comment by jorisvandenbossche** [comment]
The value of `sum` is certainly less clear, but I think `min`/`max` are nice to have?

**Comment by dsaxton** [comment]
> Just as a counterpoint do we really want to support all of these? `sum` in particular is strange to me to support on a string dtype

Agreed summing strings is a little odd, but is it worth implementing for the sake of consistency with Series of object dtype (for which this is a valid operation)?

**Comment by WillAyd** [comment]
I don’t think consistency with object dtype is a goal for the string dtype. Even for min/max I’m not sure what those mean in a lot of cases, unless the answer is to fallback to Python semantics. 

My concern is I think that as an answer conflicts with the goal of creating a native string type

**Comment by dsaxton** [comment]
> I don’t think consistency with object dtype is a goal for the string dtype. Even for min/max I’m not sure what those mean in a lot of cases, unless the answer is to fallback to Python semantics.
> 
> My concern is I think that as an answer conflicts with the goal of creating a native string type

Fair point about consistency not being important. I do think though that strings having an "order" to them is a pretty useful / natural concept (we'd probably want to allow sorting of strings, in which case we'd also want min and max)

**Comment by jorisvandenbossche** [comment]
>  Even for min/max I’m not sure what those mean in a lot of cases, unless the answer is to fallback to Python semantics.

You can sort strings, and then min/max have a rather clear meaning IMO

But it's true we certainly don't need to do this for consistency with object dtype, but because we think it is useful 
(we only need to check / fix consistency between dataframe vs series operation)

**Comment by jreback** [comment]
i would certain add operations that work now on object types - otherwise the dtypes won’t be used generally which is not great

**Comment by jorisvandenbossche** [comment]
More thoughts on adding `min`/`max` and/or `sum`. I would like to see min/max added, but care less about `sum`.

File: pandas/tests/arrays/string_/test_string.py [hunk 2]
```diff
@@ -269,3 +267,31 @@ def test_value_counts_na():
...
+@pytest.mark.parametrize("skipna", [True, False])
+def test_reduction(func):
+    strings = ["x", "y", "z"]
+    arr = pd.Series(strings, dtype="string")
+    ser = pd.Series(strings)
+
+    result = getattr(arr, func)(skipna=skipna)
+    expected = getattr(ser, func)(skipna=skipna)
+
+    assert result == expected
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/string_/test_string.py]: Instead of comparing to object dtype behavior, I would maybe rather hardcode the expected result, or use numpy for the expected behaviour

Comment (dsaxton) [reply pandas/tests/arrays/string_/test_string.py]: Still some test failures I need to look at (I think they're tests that expect methods not to be implemented). Is the general approach here valid or a bit too hacky? I suppose a better technique would just be to implement these methods for StringArray directly, but maybe do that in a future PR?

Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py]: > Is the general approach here valid or a bit too hacky? I suppose a better technique would just be to implement these methods for StringArray directly, but maybe do that in a future PR?

I think both options are fine for me (but eventually, yes, we should also fix the methods on the array class as well)

File: pandas/core/arrays/string_.py [hunk 3]
```diff
@@ -274,7 +274,16 @@ def astype(self, dtype, copy=True):
         return super().astype(dtype, copy)
 
     def _reduce(self, name, skipna=True, **kwargs):
-        raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
+        if name in ["min", "max", "sum"]:
+            na_mask = isna(self)
```
Comment (jreback) [inline pandas/core/arrays/string_.py]: the masking should be done *inside* the methods themselves, _reduce just dispatches

File: pandas/core/arrays/string_.py [hunk 4]
```diff
@@ -274,7 +274,16 @@ def astype(self, dtype, copy=True):
...
+            if not na_mask.any():
+                return getattr(self, name)(skipna=False, **kwargs)
+            elif skipna:
+                return getattr(self[~na_mask], name)(skipna=False, **kwargs)
+            else:
+                return libmissing.NA
+        else:
```
Comment (jreback) [inline pandas/core/arrays/string_.py]: you don't need the else

File: pandas/tests/arrays/string_/test_string.py [hunk 5]
```diff
@@ -269,3 +267,26 @@ def test_value_counts_na():
     result = arr.value_counts(dropna=True)
     expected = pd.Series([2, 1], index=["a", "b"], dtype="Int64")
     tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("func", ["min", "max"])
+@pytest.mark.parametrize("skipna", [True, False])
+def test_reduction(func, skipna):
```
Comment (jreback) [inline pandas/tests/arrays/string_/test_string.py]: can you test a series with nulls as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/arrays/string_/test_string.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/string_/test_string.py]: See the test below

File: pandas/core/arrays/string_.py (hunk 3 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: Should we implement these methods for StringArray in that case? The NA handling for PandasArray seems to be broken for string inputs, so it might have to get handled within each method

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: Yes, I would say don't care about PandasArray too much, and just implement the methods here on StringArray.

File: pandas/tests/arrays/string_/test_string.py (hunk 2 above)
Comment (dsaxton) [reply pandas/tests/arrays/string_/test_string.py]: I think the reason why the NA-handling wasn't working was due to an apparently long-standing bug in `nanops.nanminmax` which I think we can fix here: https://github.com/pandas-dev/pandas/issues/18588. Basically we are filling NA with infinite values when taking the min or max, but this doesn't make sense for object dtypes and an error gets raised even if `skipna` is `True`.

If we fix that by explicitly masking the missing values instead, I believe we can just use this function directly in `StringArray` methods.

File: pandas/core/arrays/string_.py (hunk 3 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: I think the reason why the NA-handling wasn't working was due to an apparently long-standing bug in nanops.nanminmax which I think we can fix here: #18588. Basically we are filling NA with infinite values when taking the min or max, but this doesn't make sense for object dtypes and an error gets raised even if skipna is True.

If we fix that by explicitly masking the missing values instead, I believe we can just use this function directly in StringArray methods.

File: pandas/tests/frame/test_apply.py [hunk 6]
```diff
@@ -1406,8 +1406,8 @@ def test_apply_datetime_tz_issue(self):
     @pytest.mark.parametrize("method", ["min", "max", "sum"])
     def test_consistency_of_aggregates_of_columns_with_missing_values(self, df, method):
         # GH 16832
-        none_in_first_column_result = getattr(df[["A", "B"]], method)()
-        none_in_second_column_result = getattr(df[["B", "A"]], method)()
+        none_in_first_column_result = getattr(df[["A", "B"]], method)().sort_index()
+        none_in_second_column_result = getattr(df[["B", "A"]], method)().sort_index()
```
Comment (dsaxton) [inline pandas/tests/frame/test_apply.py]: Previously the column with the missing value was getting dropped from the result so it only had a single row and the order didn't matter

File: pandas/core/arrays/string_.py [hunk 7]
```diff
@@ -274,8 +275,21 @@ def astype(self, dtype, copy=True):
...
         raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
 
+    def min(self, axis=None, out=None, keepdims=False, skipna=True):
+        nv.validate_min((), dict(out=out, keepdims=keepdims))
+        result = nanops.nanmin(self._ndarray, axis=axis, skipna=skipna)
+        return libmissing.NA if isna(result) else result
+
+    def max(self, axis=None, out=None, keepdims=False, skipna=True):
+        nv.validate_max((), dict(out=out, keepdims=keepdims))
+        result = nanops.nanmax(self._ndarray, axis=axis, skipna=skipna)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py]: There should be no need to explicitly pass through the axis keyword, I think

File: pandas/core/arrays/string_.py [hunk 8]
```diff
@@ -274,8 +275,21 @@ def astype(self, dtype, copy=True):
...
+            return getattr(self, name)(skipna=skipna, **kwargs)
+
         raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
 
+    def min(self, axis=None, out=None, keepdims=False, skipna=True):
+        nv.validate_min((), dict(out=out, keepdims=keepdims))
+        result = nanops.nanmin(self._ndarray, axis=axis, skipna=skipna)
+        return libmissing.NA if isna(result) else result
+
+    def max(self, axis=None, out=None, keepdims=False, skipna=True):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_.py]: If we add the numpy-compat keywords (axis, out, keepdims), we should also test this (or otherwise leave it out for now).

Also, if we keep them, would it be possible to move them into `**kwargs` ? That keeps the signature cleaner (clearer what is actually supported and not just ignored/validated)

File: pandas/core/nanops.py [hunk 9]
```diff
@@ -864,6 +866,12 @@ def reduction(
                 result.fill(np.nan)
             except (AttributeError, TypeError, ValueError):
                 result = np.nan
+        elif is_object_dtype(dtype) and values.ndim == 1 and na_mask.any():
+            # Need to explicitly mask NA values for object dtypes
+            if skipna:
+                result = getattr(values[~na_mask], meth)(axis)
```
Comment (jorisvandenbossche) [inline pandas/core/nanops.py]: This masking could also be done in the `min`/`max` functions? (as you had before?)

Or, another option might be to add a min/max function to `mask_ops.py`, similarly as I am doing for `sum` in https://github.com/pandas-dev/pandas/pull/30982 (but it should be simpler for min/max, as those don't need to handle the `min_count`)

File: pandas/tests/extension/base/reduce.py [hunk 10]
```diff
@@ -25,6 +25,12 @@ class BaseNoReduceTests(BaseReduceTests):
 
     @pytest.mark.parametrize("skipna", [True, False])
     def test_reduce_series_numeric(self, data, all_numeric_reductions, skipna):
+        if isinstance(data, pd.arrays.StringArray) and all_numeric_reductions in [
+            "min",
+            "max",
+        ]:
+            pytest.skip("These reductions are implemented")
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/reduce.py]: Can you see if you can rather update this in `test_string.py` ? It might be we now need to subclass the ReduceTests instead of NoReduceTests. 
(ideally the base tests remain dtype agnostic)

File: pandas/tests/frame/test_apply.py (hunk 6 above)
Comment (jorisvandenbossche) [inline pandas/tests/frame/test_apply.py]: What's the reason for this change?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for the update!

File: pandas/tests/frame/test_apply.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/frame/test_apply.py]: Didn't see your latest comments!

File: pandas/core/nanops.py (hunk 9 above)
Comment (dsaxton) [reply pandas/core/nanops.py]: I think a benefit of having it here is that this also fixes a bug for Series: `pd.Series(["a", np.nan]).min()` currently raises even though it shouldn't

File: pandas/core/arrays/string_.py (hunk 8 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: Ok, I was largely just mimicking the signature from PandasArray min / max

File: pandas/tests/arrays/string_/test_string.py (hunk 1 above)
Comment (dsaxton) [reply pandas/tests/arrays/string_/test_string.py]: Updated to parametrize over StringArray / Series

File: pandas/core/nanops.py (hunk 9 above)
Comment (jorisvandenbossche) [reply pandas/core/nanops.py]: Ah, that's a good point. Can you add a test for that, then? 

Now, that aside, I think longer term we still want the separate min/max in mask_ops.py, so it can also be used for the int dtypes. But that can then certainly be done for a separate PR.

File: pandas/tests/extension/base/reduce.py (hunk 10 above)
Comment (dsaxton) [reply pandas/tests/extension/base/reduce.py]: By updating in `test_string.py` do you mean adding tests using the fixtures `data` and `all_numeric_reductions`, only checking for the "correct" output (and skipping over those reductions that aren't yet implemented)?

Comment (jorisvandenbossche) [reply pandas/tests/extension/base/reduce.py]: Hmm, actually looking at the base reduction tests now: they are not really written in a way that they will pass for strings. 

But so you can copy this test to `tests/extension/test_strings.py` (and so override the base one), and then do the string-array-specific adaptation there.  It gives some duplication of the test code, but it's not long, and it clearer separation of concerns (the changes for string array are in test_string)

Comment (dsaxton) [reply pandas/tests/extension/base/reduce.py]: Ok, so we can remove the special cases for `StringArray` in `BaseNoReduceTests` without getting test failures, as long as they're handled in `TestNoReduce` in `test_string.py`? I'm not too familiar with how these particular tests actually get executed during CI

File: doc/source/whatsnew/v1.0.2.rst [hunk 11]
```diff
@@ -28,6 +28,11 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 
+**ExtensionArray**
+
+- Fixed issue where taking the minimum or maximum of a ``StringArray`` or ``Series`` with ``StringDtype`` type would raise. (:issue:`31746`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.2.rst]: does this also close #18588 can you add a note for that

Comment (jreback) [inline doc/source/whatsnew/v1.0.2.rst]: say ``.min()`` or ``.max()``

File: pandas/core/nanops.py [hunk 12]
```diff
@@ -854,6 +854,8 @@ def reduction(
         mask: Optional[np.ndarray] = None,
     ) -> Dtype:
 
+        na_mask = isna(values)
```
Comment (jreback) [inline pandas/core/nanops.py]: you should *already* have the mask (pass it in when you call this).

File: pandas/core/nanops.py [hunk 13]
```diff
@@ -864,6 +866,12 @@ def reduction(
                 result.fill(np.nan)
             except (AttributeError, TypeError, ValueError):
                 result = np.nan
+        elif is_object_dtype(dtype) and values.ndim == 1 and na_mask.any():
```
Comment (jreback) [inline pandas/core/nanops.py]: do you have a test case that fails on non ndim==1?

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.2.rst (hunk 11 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.0.2.rst]: Yes, it does close that issue

File: pandas/core/nanops.py (hunk 13 above)
Comment (dsaxton) [reply pandas/core/nanops.py]: Yes, was getting a couple test failures otherwise, I think for reductions when the entire `DataFrame` has object dtype (I can't recall which tests exactly). I figured the subsetting `values[~mask]` is only going to make sense if `values` has one dimension.

File: pandas/core/arrays/string_.py (hunk 8 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: If we remove the numpy arguments, I'm assuming we also don't need `nv.validate_min` and `nv.validate_max`?

File: doc/source/whatsnew/v1.0.2.rst [hunk 14]
```diff
@@ -30,7 +30,7 @@ Bug fixes
 
 **ExtensionArray**
 
-- Fixed issue where taking the minimum or maximum of a ``StringArray`` or ``Series`` with ``StringDtype`` type would raise. (:issue:`31746`)
+- Fixed issue where taking :meth:`min` or :meth:`max` of a ``StringArray`` or ``Series`` with ``StringDtype`` type would raise. (:issue:`31746`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.0.2.rst]:
```suggestion
- Fixed issue where taking ``min`` or ``max`` of a ``StringArray`` or ``Series`` with ``StringDtype`` type would raise. (:issue:`31746`)
```

(the min/max methods can't be linked to, because they don't exist as html page. And also if they existed, it would need to be ``` :meth:`pandas.arrays.StringArray.min` ```)

File: pandas/core/arrays/string_.py (hunk 8 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: > If we remove the numpy arguments, I'm assuming we also don't need nv.validate_min and nv.validate_max?

Yes, in that case, that isn't needed. 

Now, maybe others prefer to keep the compatibility. But I personally think we can better tackle that separately later

File: pandas/tests/extension/test_string.py [hunk 15]
```diff
@@ -77,7 +77,19 @@ class TestMissing(base.BaseMissingTests):
...
 
 class TestNoReduce(base.BaseNoReduceTests):
-    pass
+    @pytest.mark.parametrize("skipna", [True, False])
+    def test_reduce_series_numeric(self, data, all_numeric_reductions, skipna):
+        if isinstance(data, pd.arrays.StringArray) and all_numeric_reductions in [
+            "min",
+            "max",
+        ]:
+            pytest.skip("These reductions are implemented")
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_string.py]: Can you add here a comment saying that those are tested in tests/arrays/test_string.py ?

File: pandas/core/arrays/string_.py (hunk 8 above)
Comment (dsaxton) [reply pandas/core/arrays/string_.py]: That works for me, I'd also be in favor of removing the arguments if they'd never be used, but curious how others feel

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_.py]: So to be clear: they are used if you do `np.min(string_array)` (that's what this numpy compatibility is doing). However, if we support this: 1) I would move those keywords into `**kwargs` to give them less visibility (you can't use them, it's just to avoid raising an error when doing np.min(..)), 2) we should also validate axis to be None or 0, 3) we should test this.

So for those reasons, I'd rather do it in a separate PR (but you can do it here as well, it might just generate some extra discussion ;))

Comment (dsaxton) [reply pandas/core/arrays/string_.py]: Yeah I like the `kwargs` idea, and I think a separate PR would help streamline / focus the discussion a bit

File: doc/source/whatsnew/v1.0.2.rst [hunk 16]
```diff
@@ -51,4 +58,4 @@ Bug fixes
 Contributors
 ~~~~~~~~~~~~
 
-.. contributors:: v1.0.1..v1.0.2|HEAD
\ No newline at end of file
+.. contributors:: v1.0.1..v1.0.2|HEAD
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.2.rst]: what changed here?

File: pandas/core/arrays/string_.py (hunk 8 above)
Comment (jreback) [reply pandas/core/arrays/string_.py]: so we already do this for series (accept the numpy args *only* as kwargs), a followon PR would be fine

File: pandas/core/nanops.py [hunk 17]
```diff
@@ -228,7 +228,9 @@ def _maybe_get_mask(
             # Boolean data cannot contain nulls, so signal via mask being None
             return None
 
-        if skipna:
+        if skipna or is_object_dtype(values.dtype):
+            # The masking in nanminmax does not work for object dtype, so always
```
Comment (jreback) [inline pandas/core/nanops.py]: rather than do this, what exactly s the issue? 'does not work' is not very descriptive and generally we don't put comments like this, we simply fix it

File: pandas/core/nanops.py [hunk 18]
```diff
@@ -865,6 +867,17 @@ def reduction(
                 result.fill(np.nan)
             except (AttributeError, TypeError, ValueError):
                 result = np.nan
+        elif (
+            is_object_dtype(dtype)
+            and values.ndim == 1
+            and mask is not None
+            and mask.any()
+        ):
+            # Need to explicitly mask NA values for object dtypes
```
Comment (jreback) [inline pandas/core/nanops.py]: why?

File: pandas/core/nanops.py [hunk 19]
```diff
@@ -865,6 +867,17 @@ def reduction(
                 result.fill(np.nan)
             except (AttributeError, TypeError, ValueError):
                 result = np.nan
+        elif (
+            is_object_dtype(dtype)
```
Comment (jreback) [inline pandas/core/nanops.py]: why do you need all of these condtions? this is complicated

**Review by jreback — CHANGES_REQUESTED** [review]
more comments

File: doc/source/whatsnew/v1.0.2.rst (hunk 16 above)
Comment (dsaxton) [reply doc/source/whatsnew/v1.0.2.rst]: I think the original file lacked a newline at the end of file so I may have added it to placate git. Can remove if we want to clear up the diff.

File: pandas/core/nanops.py (hunk 17 above)
Comment (dsaxton) [reply pandas/core/nanops.py]: So what nanops.nanminmax appears to do when taking the min or max in the presence of missing values is to fill them with an appropriate infinite number that has the effect of ignoring those missing values (if we're taking the minimum replace with positive infinity, if we're taking the max replace with negative infinity). The problem is that this makes no sense for strings (there is as far as I know no "infinite string"), and that's why we get the error about comparing floats (infinity) and strings. The easiest workaround seems to be to mask them out instead.

To make things more complicated the `_get_values` function in `nanminmax` apparently doesn't bother to calculate a missing value mask when `skipna` is `False` because it's relying on the trick above working. Since it won't I'm making sure that we always get a mask for object dtypes.

File: doc/source/whatsnew/v1.0.2.rst (hunk 16 above)
Comment (jreback) [reply doc/source/whatsnew/v1.0.2.rst]: no it’s fine

File: pandas/core/nanops.py (hunk 19 above)
Comment (dsaxton) [reply pandas/core/nanops.py]: Only looking at objects is for the reason above, `values.ndim == 1` is because I think we can get here even if `values` is not vector-valued, which is more or less what we're assuming when we just mask them out (if we don't include this we can get test failures when we have an all-object `DataFrame`), `mask.any()` is because this function already works if no values are missing so there's no reason to do masking, and `mask is not None` is to please `mypy` (we've already guaranteed that `mask` isn't `None` for objects above but `mypy` doesn't know this).

I can try to be a bit more explicit in the comments if that would be helpful.

File: pandas/core/nanops.py (hunk 17 above)
Comment (jreback) [reply pandas/core/nanops.py]: ok, then actually put this expl there. yes this might need a refactor for object type (which is out of scope for this PR) or at least backporting. so separate this out to another PR as its not trivially fixed here.

**Comment by jreback** [comment]
moving this off 1.0.2 as this has raised some non-trivial to solve issues. if you want a bug fix out of this ok, but then this needs to decouple the issues.

File: pandas/core/nanops.py (hunk 17 above)
Comment (jorisvandenbossche) [reply pandas/core/nanops.py]: Similarly as https://github.com/pandas-dev/pandas/pull/30982, I would personally rather have a separate implementation for one using masks instead of the filling logic of the nanops (with sharing helper functions where appropriate), instead of trying to fit it into the existing nanops code (which gives those special cases / complex if checks like the one below)

Comment (dsaxton) [reply pandas/core/nanops.py]: @jorisvandenbossche Fair point for the string dtype, although I think some kind of logic like this would be necessary to fix `min / max` for object strings.

Comment (jorisvandenbossche) [reply pandas/core/nanops.py]: > some kind of logic like this would be necessary to fix min / max for object strings.

Yep, indeed, that's what you mentioned as argument before as well for doing it in nanops, so the non-extension array object dtype would benefit as well. For this, we could also add a check for object dtype, and then calculate the mask and use the masked op implementation.

Comment (dsaxton) [reply pandas/core/nanops.py]: Do you think it might also be worth trying to refactor nanminmax in nanops to use masking in general instead of the current filling approach (from what I could tell this was only really needed for arrays with more than one dimension)?

Comment (jreback) [reply pandas/core/nanops.py]: yes

File: doc/source/whatsnew/v1.0.2.rst [hunk 20]
```diff
@@ -59,6 +59,10 @@ Previously indexing with a nullable Boolean array containing ``NA`` would raise
 Bug fixes
 ~~~~~~~~~
 
+**ExtensionArray**
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.2.rst]: this is likely too invasive for 1.02, move to 1.1

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by dsaxton** [comment]
Going to close this for now since it looks to be quite a bit larger than I originally thought (seems like fixing nanmin / nanmax might be the right approach).

I'm not sure how best to implement this, but another idea might be to fill with an arbitrary non-NaN value from each array along the given axis before taking the min or max, then I think it should work for any dtype (although might be a little slower).
